frontend/lib/resolution/Resolver.cpp

```diff
@@ -2563,14 +2563,69 @@ Resolver::adjustTypesForOutFormals(const CallInfo& ci,
   }
 }
 
-void Resolver::resolveTupleUnpackAssign(ResolvedExpression& r,
-                                        const uast::AstNode* astForErr,
-                                        const Tuple* lhsTuple,
-                                        QualifiedType lhsType,
-                                        QualifiedType rhsType) {
-  // Check that lhsType = rhsType can work
-
-  if (!lhsType.hasTypePtr()) {
+static bool
+isValidAssignmentTarget(const ID& id, const ResolvedExpression& re) {
+  // If the kind is a reference then it must point to an assignable name.
+  if (re.type().isRef()) return true;
+  // It's a non-ref returning call, so it returns a temporary value.
+  if (re.mostSpecific().foundCandidates()) return false;
```

Comment:
How does this play with return intent overloading?

In fact, all of this handling feels like it belongs in the various `VarScopeVisitor` passes that handle refness etc.

Reply:
Shoot, I thought I removed this already...

I agree that this 'is valid assignment' check shouldn't be handled quite so manually right here. Best to leave that up to const-checking, I think.